### PR TITLE
feat(ssh): allow to use pubkey auth

### PR DESCRIPTION
**Pull request by alexandrebrg**
### What changes are being made and why?

:wave: 

SSH plugin command was missing authentication with asymmetric keys. I added an option to determine which option will be used in the SSH command (`PASSWORD` vs `PUBLIC_KEY`), with the ability to give a private key and a passphrase for it. 

Moreover, **JSch** has been replaced by this [fork](https://github.com/mwiede/jsch). It's a drop-in replacement, with enhanced capabilities, especially in cryptography. Original version was only able to run `ssh-rsa` keys, which is a method deprecated nowadays in OpenSSL, and even more, it's disabled by default. (see [here](https://stackoverflow.com/questions/72743823/public-key-authentication-fails-with-jsch-but-work-with-openssh-with-the-same-ke)).

Based on my light research, this lib wasn't used by any other component, meaning it shouldn't impact other plugins. A little note here to talk about available algorithms, this lib will load various (most commons) algo depending on the java version is it built with ([doc](https://github.com/mwiede/jsch?tab=readme-ov-file#are-ssh-ed25519-ssh-ed448-curve25519-sha256-curve448-sha512--chacha20-poly1305opensshcom-supported)). If I remember well, Kestra must be built & run with at least java 17, so it should be good :D

### How the changes have been QAed?

I'm running a simple docker-compose as described [here](https://gist.github.com/alexandrebrg/073434b917d7e1a428791e546eba1ca9), with the below flow and this additional docker compose service:

```yaml
services:
  openssh-server:
    image: lscr.io/linuxserver/openssh-server:latest
    container_name: openssh-server
    hostname: openssh-server #optional
    environment:
      PUBLIC_KEY: ${ED255_19_PUB_KEY_HERE}
      PUID: 1000
      PGID: 1000
      TZ: Etc/UTC
      LOG_STDOUT: true
      PASSWORD_ACCESS: true
      USER_PASSWORD: password
    ports:
      - 2222:2222
    restart: unless-stopped
```

```yaml
# Run the plugin with the two methods: password, public_key
id: hello-world
namespace: company.team
tasks:
  - id: "command"
    type: "io.kestra.plugin.fs.ssh.Command"
    host: openssh-server
    port: "2222"
    authMethod: PUBLIC_KEY
    username: linuxserver.io
    privateKey: ${ED_KEY}
    commands:
      - "touch i_was_here"
  - id: "command-password"
    type: "io.kestra.plugin.fs.ssh.Command"
    host: openssh-server
    port: "2222"
    authMethod: PASSWORD
    username: linuxserver.io
    password: password
    commands:
      - "touch i_was_here"
```

---

### Setup Instructions

1. Generate key `ssh-keygen -t ed255_19 -f keypair -P "" -N "" `
2. Create a simple SSH server (see above)
3. Create flow like above
4. Run it, enjoy

### CODE REVIEW

File: src/main/java/io/kestra/plugin/fs/ssh/Command.java [hunk 1]
```diff
@@ -46,22 +45,43 @@
...
+    private AuthMethod authMethod;
+
+    // Password Auth method
     private String username;
-    private String password;
+    private Optional<String> password;
+
+    // PubKey Auth method
+    private Optional<String> privateKey;
+    private Optional<String> privateKeyPassphrase;
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/fs/ssh/Command.java:83]: We usually don't use Optional but check for not null instead.
I agree that Optional is convenient but please remove them for consistency with other plugins.

File: src/main/java/io/kestra/plugin/fs/ssh/Command.java [hunk 2]
```diff
@@ -46,22 +45,43 @@
...
             }
         )
     }
 )
-public class Command extends Task implements AbstractVfsInterface, RunnableTask<Command.ScriptOutput> {
+public class Command extends Task implements SshInterface, RunnableTask<Command.ScriptOutput> {
     private static final long SLEEP_DELAY_MS = 25L;
 
     private String host;
+    private AuthMethod authMethod;
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/fs/ssh/Command.java:75]: You should init the value with the default auth method, something like that:

```java
@Builder.Default
private AuthMethod authMethod = AuthMethod.PASSWORD;
```

File: src/main/java/io/kestra/plugin/fs/ssh/Command.java [hunk 3]
```diff
@@ -86,11 +106,23 @@ public class Command extends Task implements AbstractVfsInterface, RunnableTask<
...
     public Command.ScriptOutput run(RunContext runContext) throws Exception {
+        JSch jsch = null;
         Session session = null;
         ChannelExec channel = null;
         LogThread stdOut = null;
         LogThread stdErr = null;
 
+        if (authMethod == AuthMethod.PASSWORD) {
+            if (password == null || password.get().isEmpty()) {
+                throw new RuntimeException("Password is necessary for given SSH auth method: " + AuthMethod.PASSWORD);
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/fs/ssh/Command.java:117]:
```suggestion
                throw new IllegalArgumentException("Password is necessary for given SSH auth method: " + AuthMethod.PASSWORD);
```

File: src/main/java/io/kestra/plugin/fs/ssh/Command.java [hunk 4]
```diff
@@ -86,11 +106,23 @@ public class Command extends Task implements AbstractVfsInterface, RunnableTask<
...
+            }
+        }
+        else if (authMethod == AuthMethod.PUBLIC_KEY) {
+            if (privateKey == null || privateKey.get().isEmpty()) {
+                throw new RuntimeException("Private key is necessary for given SSH auth method: " + AuthMethod.PASSWORD);
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/fs/ssh/Command.java:122]:
```suggestion
                throw new IllegalArgumentException("Private key is necessary for given SSH auth method: " + AuthMethod.PASSWORD);
```

File: src/main/java/io/kestra/plugin/fs/ssh/SshInterface.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        title = "Port of the remote server"
+    )
+    @PluginProperty(dynamic = true)
+    String getPort();
+
+    @Schema(
+        title = "Authentication method"
+    )
+    @PluginProperty(dynamic = true)
+    AuthMethod getAuthMethod();
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/fs/ssh/SshInterface.java:28]:
```suggestion
    @NotNull
    AuthMethod getAuthMethod();
```

**Review by loicmathieu — APPROVED** [review]
Hi, thanks a lot.
Overall it looks good, I suggest some small code changes to make it coherent with the way we declare plugin properties.

**Comment by alexandrebrg** [comment]
> Hi, thanks a lot. Overall it looks good, I suggest some small code changes to make it coherent with the way we declare plugin properties.

Thanks for the review, I applied the changes

**Review by loicmathieu — APPROVED** [review]
LGTM, thanks a lot.

**Comment by alexandrebrg** [comment]
@loicmathieu Tests should now be passing properly :)

**Review by loicmathieu — APPROVED** [review]
LGTM thanks a lot

**Pull request merged**